### PR TITLE
Enable Exposure of Survey Fields

### LINKED ISSUE #55: [BUG] REDCapTidieR doesn't return survey timestamp fields
Amazing, elegant work 💪  I have been battling with REDCap's 'block matrix' data format written my own much less elegant conversion scripts. Thanks so much for sharing with the community.

# Enhancment Idea
Unless have overlooked something, currently the CRAN version of REDCapTidieR doesn't expose the survey timestamp field.

# Proposed Solution
Expose the 'survey_timestamp' field for all instruments of type survey similar to 'form_status_complete'.

### PR DESCRIPTION
**Pull request by rsh52**
# Description
This PR makes use of `REDCapR::redcap_read_oneshot()`'s `export_survey_fields` argument to allow for default survey fields in supertibble extracts (`redcap_survey_identifier` and `[form_name]_timestamp`).

# Proposed Changes 
List changes below in bullet format:

- Include `export_survey_fields` argument in `read_redcap_tidy`
- Update `distill_*` functions to include passive, but literal, `match`ing and identification of survey timestamp fields
- Write applicable tests
- Update mocks to include new mock with export set to `TRUE`

## Other Notes:
Surveys work mostly the same as any other instrument in the database, but with a few caveats. For one, there is no real way to identify a "survey" instrument in the metadata.

The `redcap_survey_identifier` field behavior seems to always comes out with the export, and for anonymous surveys this will always be `NULL`. You can only set respondent identifiers (i.e. a non-anonymous survey) by making a survey instrument out of the same instrument that contains the `record_id`. I tested this with the Classic test database and this was the only way to get identifiers enabled. Once identifiers are enabled, they only get associated with either non-repeat observations or the first instance of a repeat observation. This can be a little messy, but I think the best avenue here is to present the user with the information "as is." 

There is a lot of intricacy to how surveys get sent from inside of the REDCap interface using the Distribution Tool, but I don't think they apply to our package. Outside of these two variables, I don't think `REDCapTidieR` is responsible for other survey behavior.

### Issue Addressed
Addresses the main request in #55, but not the subsequent "never invited" survey (I'm not sure this is in scope for us).

**Closes #55**

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [x] New tests that make API calls use `httptest::with_mock_api` and any new mocks were added to `tests/testthat/fixtures/create_httptest_mocks.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

### CODE REVIEW

**Comment by skadauke** [comment]
Great start. I think it would be better to just always export the survey timestamps. I don't think we gain anything from leaving this out in terms of performance or memory use.

**Comment by rsh52** [comment]
> Great start. I think it would be better to just always export the survey timestamps. I don't think we gain anything from leaving this out in terms of performance or memory use.

Sounds good, I don't feel strongly one way or another since the output doesn't really change. Only argument is `REDCapR` has it set to `FALSE` as default so maybe we should follow suit?

File: R/clean_redcap_long.R [hunk 1]
```diff
@@ -155,9 +155,15 @@ distill_nonrepeat_table_long <- function(
   # Below necessary to remove descriptive text fields
   # and to add column to indicate that form is completed
   my_fields <- db_data_long %>%
-    select(all_of(my_fields), paste0(my_form, "_complete")) %>%
+    select(all_of(my_fields),
+           matches(paste0("^",my_form, "_timestamp$")),
+           paste0(my_form, "_complete")) %>%
```
Comment (skadauke) [inline R/clean_redcap_long.R:160]: Can you delete ll. 163-165 and add after 159

`matches("^redcap_survey_identifier$"),`

?

File: R/clean_redcap_long.R [hunk 2]
```diff
@@ -240,9 +246,15 @@ distill_repeat_table_long <- function(
...
     names()
 
+  if (paste0(my_form, "_timestamp") %in% names(db_data_long)) {
```
Comment (skadauke) [inline R/clean_redcap_long.R:254]: See above comment

**Comment by skadauke** [comment]
> > Great start. I think it would be better to just always export the survey timestamps. I don't think we gain anything from leaving this out in terms of performance or memory use.
> 
> Sounds good, I don't feel strongly one way or another since the output doesn't really change. Only argument is `REDCapR` has it set to `FALSE` as default so maybe we should follow suit?

I think the only good reason to follow suit here would be unacceptable performance degradation. Additional columns hurt little and can help a lot.

File: R/clean_redcap_long.R (hunk 1 above)
Comment (rsh52) [reply R/clean_redcap_long.R:160]: Yes, good call!

**Comment by rsh52** [comment]
@ezraporter and I will figure out this vignette/mocks issue on Monday 😓

File: R/clean_redcap.R [hunk 3]
```diff
@@ -144,7 +145,10 @@ distill_nonrepeat_table <- function(
   # Below necessary to remove descriptive text fields
   # and to add column to indicate that form is completed
   my_fields <- db_data %>%
-    select(all_of(my_fields), paste0(my_form, "_complete")) %>%
+    select(all_of(my_fields),
+           matches(paste0("^",my_form, "_timestamp$")),
+           matches("^redcap_survey_identifier$"),
```
Comment (ezraporter) [inline R/clean_redcap.R:150]: Would this work? I think it better represents the intention of "these fields if they're in the data"
```suggestion
           anyof(c(paste0(my_form, "_timestamp"), "redcap_survey_identifier")),
```

File: R/clean_redcap.R [hunk 4]
```diff
@@ -202,7 +206,10 @@ distill_repeat_table <- function(
...
```
Comment (ezraporter) [inline R/clean_redcap.R:211]:
```suggestion
           anyof(c(paste0(my_form, "_timestamp"), "redcap_survey_identifier")),
```

File: R/clean_redcap.R [hunk 5]
```diff
@@ -109,7 +110,7 @@ clean_redcap <- function(
 #' \code{REDCapR::redcap_metadata_read()$data}
 #'
 #' @importFrom dplyr filter pull select relocate rename
-#' @importFrom tidyselect all_of everything starts_with
+#' @importFrom tidyselect all_of everything starts_with matches
```
Comment (ezraporter) [inline R/clean_redcap.R:113]: If you adopt my suggestion below:
```suggestion
#' @importFrom tidyselect all_of everything starts_with any_of
```

File: R/clean_redcap.R [hunk 6]
```diff
@@ -177,7 +181,7 @@ distill_nonrepeat_table <- function(
...
```
Comment (ezraporter) [inline R/clean_redcap.R:184]:
```suggestion
#' @importFrom tidyselect all_of everything starts_with any_of
```

File: R/clean_redcap_long.R [hunk 7]
```diff
@@ -119,7 +119,7 @@ clean_redcap_long <- function(
 #' events/arms
 #'
 #' @importFrom dplyr filter pull select relocate rename
-#' @importFrom tidyselect all_of everything
+#' @importFrom tidyselect all_of everything matches
```
Comment (ezraporter) [inline R/clean_redcap_long.R:122]:
```suggestion
#' @importFrom tidyselect all_of everything any_of
```

File: R/clean_redcap_long.R [hunk 8]
```diff
@@ -155,7 +155,10 @@ distill_nonrepeat_table_long <- function(
   # Below necessary to remove descriptive text fields
   # and to add column to indicate that form is completed
   my_fields <- db_data_long %>%
-    select(all_of(my_fields), paste0(my_form, "_complete")) %>%
+    select(all_of(my_fields),
+           matches(paste0("^",my_form, "_timestamp$")),
+           matches("^redcap_survey_identifier$"),
```
Comment (ezraporter) [inline R/clean_redcap_long.R:160]:
```suggestion
           anyof(c(paste0(my_form, "_timestamp"), "redcap_survey_identifier")),
```

File: R/clean_redcap_long.R [hunk 9]
```diff
@@ -214,7 +217,7 @@ distill_nonrepeat_table_long <- function(
 #' events/arms
 #'
 #' @importFrom dplyr filter pull select relocate rename
-#' @importFrom tidyselect all_of everything
+#' @importFrom tidyselect all_of everything matches
```
Comment (ezraporter) [inline R/clean_redcap_long.R:220]:
```suggestion
#' @importFrom tidyselect all_of everything any_of
```

File: R/clean_redcap_long.R [hunk 10]
```diff
@@ -240,7 +243,10 @@ distill_repeat_table_long <- function(
   # Below necessary to remove descriptive text fields
   # and to add column to indicate that form is completed
   my_fields <- db_data_long %>%
-    select(all_of(my_fields), paste0(my_form, "_complete")) %>%
+    select(all_of(my_fields),
+           matches(paste0("^",my_form, "_timestamp$")),
+           matches("^redcap_survey_identifier$"),
```
Comment (ezraporter) [inline R/clean_redcap_long.R:248]:
```suggestion
           anyof(c(paste0(my_form, "_timestamp"), "redcap_survey_identifier")),
```

**Review by ezraporter — CHANGES_REQUESTED** [review]
Looks pretty good and I like your test cases. See my one suggestion about `any_of()` vs. `match()`. Pretty much all of my comments are flagging the places where that's relevant.

**Comment by skadauke** [comment]
I like `any_of`!

**Comment by rsh52** [comment]
Sure, I'll implement now sounds good to me.

**Review by ezraporter — APPROVED** [review]
🚀
